1.2.5 didn't support activated virtualenv

artburkart opened the pull request:
I found [this answer](http://stackoverflow.com/questions/21698004/python-behave-integration-in-setuptools-setup-py#answer-22901706) on Stackoverflow and realized `setuptools_behave` doesn't work in a virtualenv environment. My solution only addresses a one-off and thus is not ideal, but I wanted to get the ball rolling on this, so others could maybe point out a better direction.

[Maybe something like this is better?](https://github.com/python-standard-library-examples/subprocess.python.standard-library.examples/blob/b96a5b8896902544a0781a96c75dcab8d52fd2df/python/subprocess/Popen(args%2C%20bufsize%3D0%2C%20executable%3DNone%2C%20stdin%3DNone%2C%20stdout%3DNone%2C%20stderr%3DNone%2C%20preexec_fn%3DNone%2C%20close_fds%3DFalse%2C%20shell%3DFalse%2C%20cwd%3DNone%2C%20env%3DNone%2C%20universal_newlines%3DFalse%2C%20startupinfo%3DNone%2C%20creationflags%3D0)/Popen(%5B%22which%22%5D)/distutils.spawn.find_executable())

```python
from distutils.spawn import find_executable
import os.path

binpath = os.path.dirname(find_executable('python'))
```

Of course, I'm not sure how the latter would address things like python3 in a non-virtualenv environment.

Any thoughts? Cheers!

artburkart commented:
No interest in this one?

jenisys commented:
@artburkart 
Before I fix something, I need to know why this fix is needed.
I cannot reproduce your results. I works fine for me:

```sh
python setup.py behave_test -f progress3 --args features/

# HINT: Ensure behave works correctly by using (before you use: python setup.py behave_test)
behave -f progress3 features/
```

NOTE:
Runs fine for me in an other non-behave project, with a virtual environment for Python2.7 and the current HEAD of the behave repository or latest stable version (1.2.6.dev0, 1.2.5) installed in it. The virtualenv is activated.

Why would I need a ``binpath`` to the python executable? 
When you already activated the virtual environment, its python executable is already accessible via the ``PATH`` environment variable.

MORE INFO NEEDED:
* Which Python version are you using ?
* On which platform ? (like: Windows7, Linux, OSX 10.11, ...)
* Which behave version (1.2.5 or Git repo HEAD 1.2.6.dev0) ?
* Which virtualenv version are using (with which pip and setuptools version) ?

EXAMPLE:
```sh
$ pip list
...
behave (1.2.6.dev0)
...
pip (8.1.2)
...
setuptools (24.0.3)
...
virtualenv (15.0.2)
wheel (0.29.0)
```

jenisys commented:
NOTE:
If you already within python, ``sys.executable`` tells you the path to the Python executable that you are currently using.

artburkart commented:
Sorry for the brevity of my PR. I don't tend to put too much effort into OS projects unless they have a clear contributors guide.

Here is all the info needed to reproduce:

- python version

```bash
$ python --version
Python 3.5.1+
```

- pip version

```bash
$ pip --version
pip 8.1.2 from /usr/local/lib/python3.5/dist-packages (python 3.5)
```

- OS
```bash
$ uname -a
Linux vendetta3 4.4.0-28-generic #47-Ubuntu SMP Fri Jun 24 10:09:13 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

- requirements.yml

```
behave==1.2.5
parse==1.6.6
parse-type==0.3.4
PyHamcrest==1.9.0
six==1.10.0
```

installed pip packages

```bash
$ pip list

behave (1.2.5)
parse (1.6.6)
parse-type (0.3.4)
pip (8.1.2)
PyHamcrest (1.9.0)
setuptools (25.1.0)
six (1.10.0)
wheel (0.29.0)
```

- virtualenv version

```bash
$ virtualenv --version
15.0.2
```

Structure of repo:

```
app_code
├── app_code
│   ├── __init__.py
│   └── tests
│       ├── features
│       │   ├── example.feature
│       │   └── steps
│       │       └── example_steps.py
│       └── __init__.py
```

Error message I get:

```bash
$ python setup.py behave_test -f progress3 --args app_code/tests/features/

running behave_test
Use PYTHONPATH=/home/artburkart/Code/app_code
CMDLINE: behave  --format=progress3 app_code/tests/features/
/home/artburkart/Code/app_code/app_code/venv/bin/python: can't open file 'behave': [Errno 2] No such file or directory
FAILED
```

*HINT: I ensured behave works without setup.py*

```
$ which behave
/home/artburkart/Code/app_code/venv/bin/behave

$ behave -f progress3 app_code/tests/features/

$ behave -f progress3 app_code/tests/features/
Reading deployment files    # app_code/tests/features/example.feature
  Execute code before downloading artifact  U


Failing scenarios:
  app_code/tests/features/example.feature:3  Execute code before downloading artifact

0 features passed, 1 failed, 0 skipped
0 scenarios passed, 1 failed, 0 skipped
0 steps passed, 0 failed, 0 skipped, 3 undefined
Took 0m0.000s

You can implement step definitions for undefined steps with these snippets:

@given(u'a deployment has been triggered')
def step_impl(context):
    raise NotImplementedError(u'STEP: Given a deployment has been triggered')

@when(u'app_code pulls down the deployment config file')
def step_impl(context):
    raise NotImplementedError(u'STEP: When app_code pulls down the deployment config file')

@then(u'it should execute code associated with the before download step')
def step_impl(context):
    raise NotImplementedError(u'STEP: Then it should execute code associated with the before download step')
```

How did you create the virtual environment?

```
$ virtualenv -p `which python3` venv
$ . venv/bin/activate
$ python3 setup.py install
```

re: 
> Why would I need a binpath to the python executable? 

This is exactly what I'm confused about. When I print the `$PATH` at the time of execution, it is thus:

```
/home/artburkart/Code/app_code/app_code/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

---

After doing some further research, I've concluded it's a bug in `1.2.5`. The difference in code being:

- 1.2.5

```python
    def behave(self, path):
        behave = os.path.join("bin", "behave")
        if not os.path.exists(behave):
            behave = "behave"
        cmd_options = ""
        if self.tags:
            cmd_options = "--tags=" + " --tags=".join(self.tags)
        if self.dry_run:
            cmd_options += " --dry-run"
        cmd_options += " --format=%s %s" % (self.format, path)
        self.announce("CMDLINE: %s %s" % (behave, cmd_options), level=3)
        return subprocess.call([sys.executable, behave] + shlex.split(cmd_options))
```

- git

```python
    def behave(self, path):
        behave = os.path.join("bin", "behave")
        if not os.path.exists(behave):
            # -- ALTERNATIVE: USE: behave script: behave = "behave"
            # -- USE: behave module (main)
            behave = "-m behave"
        cmd_options = ""
        if self.tags:
            cmd_options = "--tags=" + " --tags=".join(self.tags)
        if self.dry_run:
            cmd_options += " --dry-run"
        cmd_options += " --format=%s %s" % (self.format, path)
        self.announce("CMDLINE: python %s %s" % (behave, cmd_options), level=3)
        behave_cmd = shlex.split(behave)
        return subprocess.call([sys.executable] + behave_cmd + shlex.split(cmd_options))
```

I imagine it's the `behave_cmd` that fixes it. 

I guess I'll pull the code from github for now.

jenisys commented:
DUPLICATES: #310
Already fixed in behave-1.2.6.dev0
Therefore, check it w/ the current head of the Git repository.

artburkart commented:
This was addressed over a year ago and has bitten other people.
Therefore, maybe it's time to publish the latest rendition of the module.

aellwein commented:
+1 for release it in PyPi

system7ltd commented:
I usually work around this problem by creating a subclass of behave_test and overriding behave_test.behave().  It has been almost two years since version 1.2.5 was released.  Any reason not to release version 1.2.6 ASAP?
